RTP stream timestamps

marchaase opened the pull request:
#485 

I added rtp timestamps to the streaming plugin.  It adds two variables to the janus_streaming_mountpoint struct: last_received_video and last_received_audio.  It uses janus_get_monotonic_time to timestamp when an RTP packet has arrived and updates the struct's variable.  In the message handling for the 'list' command it outputs the timestamps if it's an RTP stream and a 'now' json member which is the current monotonic time.  This allows a client to calculate the time since the last packet (audio or video) has arrived and decide if the stream has timed out/stopped sending packets.

ploxiln commented:
It might be a nicer api/interface if you calculate the age in milliseconds (`(now - last_recieved)/1000000`) In this case the json properties would be something like "audio_age_ms" and "video_age_ms", and no "now" element.

marchaase commented:
I do like that idea - that way the json member is self descriptive. An
integer is precise enough in ms for this purpose.
On Mar 15, 2016 6:17 PM, "Pierce Lopez" <notifications@github.com> wrote:

> It might be a nicer api/interface if you calculate the age in milliseconds
> ((now - last_recieved)/1000000) In this case the json properties would be
> something like "audio_age_ms" and "video_age_ms", and no "now" element.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/meetecho/janus-gateway/pull/486#issuecomment-197094928>
>

lminiero commented:
I'd move the timing info to the ```janus_streaming_rtp_source``` struct instead, since they only make sense for RTP based feeds.

lminiero commented:
As a side note, if you haven't signed our [CLA](https://cla.conf.meetecho.com/janus/) yet, please do so before we can merge your contribution.

marchaase commented:
Signed the CLA.

Implemented ms API facing changes as @ploxiln suggested.
Moved last_received memebrs to rtp struct.
Also changed the API to only output the age in ms if that part of the stream is enabled.

Changing the location of the last_received_* members also means they aren't in the mutex protected section.  There is not mutex for the rtp struct, so I didn't add one for now.  @lminiero it sounds like you are of the mind that writing to a gint64 in one thread and reading the gint64 in another thread shouldn't be an issue?  I think 64-bit values are atomic on a 64-bit architecture if they are memory aligned, but they are not on a 32-bit arch.  Probably not that bad of a problem since it's a monotonic counter - there is one possible race condition when the carry over from the 32nd bit to the 33rd bit occurs and if the read happens right there on a 32-bit arch.

lminiero commented:
You're right that this can happen, but I think it's unlikely enough that we can ignore that. Anyway, if you feel more comfortable with having a lock anyway, I think you can use the same lock you were using before, that is the lock of the mountpoint itself: after all, the source is a property of the mountpoint.

lminiero commented:
@marchaase as a side note, are you sure you signed the CLA? I can't find you in the list, vould you try again? Thanks!

lminiero commented on plugins/janus_streaming.c:
```diff
@@ -1015,6 +1017,14 @@ struct janus_plugin_result *janus_streaming_handle_message(janus_plugin_session
+				if ( source->audio_fd != -1 )
+					json_object_set_new(ml, "audio_age_ms", json_integer((now - source->last_received_audio) / 1000));
+				if ( source->video_fd != -1 )
+					json_object_set_new(ml, "video_age_ms", json_integer((now - source->last_received_video) / 1000));
```
Just for our coding conventions, could you remove those extra spaces in the two ifs?

    if(source->audio_fd != -1)

For the rest, the patch looks fine to me and, apart from the considerations on whether or not we need locking (something we can fix along the way, if needed) we can merge it. Thanks for working on this!

lminiero replied:
Nevermind that, it's such a minor fix that I can do this myself after we merge. Thanks!

marchaase commented:
Cool, sounds good Lorenzo.  Hard to force yourself into a different coding style sometimes =)

I really like the janus-gateway software, so I'm glad that I'm able to contribute what little small parts I can.

And the locking can be addressed easily if needed later, just got to move that update into the mutex protected section at the bottom of the video/audio if branches.
